// compose-area/compose_area.js
/* tslint:disable */
import * as wasm from './compose_area_bg';

const heap = new Array(32);

heap.fill(undefined);

heap.push(undefined, null, true, false);

let stack_pointer = 32;

function addBorrowedObject(obj) {
    if (stack_pointer == 1) throw new Error('out of js stack');
    heap[--stack_pointer] = obj;
    return stack_pointer;
}
/**
* Find the current caret position. If it cannot be determined, the position
* (0, 0) will be returned.
*
* Note: When getting the window selection, it will be relative to an anchor
*       node. If the anchor node is an element node, then the offset is
*       referring to the number of child elements. If the anchor node is a
*       text node, then the offset is referring to the codepoints.
*
* TODO: Return a tuple once
* https://github.com/rustwasm/wasm-bindgen/issues/122 is resolved!
*
* TODO: Make this nullable once
* https://github.com/rustwasm/wasm-bindgen/issues/1252 is resolved!
* @param {any} arg0
* @returns {CaretPosition}
*/
export function get_caret_position(arg0) {
    try {
        return CaretPosition.__wrap(wasm.get_caret_position(addBorrowedObject(arg0)));

    } finally {
        heap[stack_pointer++] = undefined;

    }

}

let cachedTextDecoder = new TextDecoder('utf-8');

let cachegetUint8Memory = null;
function getUint8Memory() {
    if (cachegetUint8Memory === null || cachegetUint8Memory.buffer !== wasm.memory.buffer) {
        cachegetUint8Memory = new Uint8Array(wasm.memory.buffer);
    }
    return cachegetUint8Memory;
}

function getStringFromWasm(ptr, len) {
    return cachedTextDecoder.decode(getUint8Memory().subarray(ptr, ptr + len));
}

let cachedGlobalArgumentPtr = null;
function globalArgumentPtr() {
    if (cachedGlobalArgumentPtr === null) {
        cachedGlobalArgumentPtr = wasm.__wbindgen_global_argument_ptr();
    }
    return cachedGlobalArgumentPtr;
}

let cachegetUint32Memory = null;
function getUint32Memory() {
    if (cachegetUint32Memory === null || cachegetUint32Memory.buffer !== wasm.memory.buffer) {
        cachegetUint32Memory = new Uint32Array(wasm.memory.buffer);
    }
    return cachegetUint32Memory;
}
/**
* Process a DOM node recursively and extract text.
*
* Convert elements like images to alt text.
* @param {any} arg0
* @param {boolean} arg1
* @returns {string}
*/
export function extract_text(arg0, arg1) {
    const retptr = globalArgumentPtr();
    try {
        wasm.extract_text(retptr, addBorrowedObject(arg0), arg1);
        const mem = getUint32Memory();
        const rustptr = mem[retptr / 4];
        const rustlen = mem[retptr / 4 + 1];

        const realRet = getStringFromWasm(rustptr, rustlen).slice();
        wasm.__wbindgen_free(rustptr, rustlen * 1);
        return realRet;


    } finally {
        heap[stack_pointer++] = undefined;

    }

}

let cachedTextEncoder = new TextEncoder('utf-8');

let WASM_VECTOR_LEN = 0;

function passStringToWasm(arg) {

    const buf = cachedTextEncoder.encode(arg);
    const ptr = wasm.__wbindgen_malloc(buf.length);
    getUint8Memory().set(buf, ptr);
    WASM_VECTOR_LEN = buf.length;
    return ptr;
}
/**
* Initialize a new compose area wrapper with the specified `id`.
* @param {string} arg0
* @returns {ComposeArea}
*/
export function bind_to(arg0) {
    const ptr0 = passStringToWasm(arg0);
    const len0 = WASM_VECTOR_LEN;
    try {
        return ComposeArea.__wrap(wasm.bind_to(ptr0, len0));

    } finally {
        wasm.__wbindgen_free(ptr0, len0 * 1);

    }

}

export function __wbg_error_cc95a3d302735ca3(arg0, arg1) {
    let varg0 = getStringFromWasm(arg0, arg1);

    varg0 = varg0.slice();
    wasm.__wbindgen_free(arg0, arg1 * 1);

    console.error(varg0);
}

function getObject(idx) { return heap[idx]; }

let heap_next = heap.length;

function addHeapObject(obj) {
    if (heap_next === heap.length) heap.push(heap.length + 1);
    const idx = heap_next;
    heap_next = heap[idx];

    heap[idx] = obj;
    return idx;
}

function handleError(exnptr, e) {
    const view = getUint32Memory();
    view[exnptr / 4] = 1;
    view[exnptr / 4 + 1] = addHeapObject(e);
}

export function __widl_f_replace_with_with_node_1_CharacterData(arg0, arg1, exnptr) {
    try {
        getObject(arg0).replaceWith(getObject(arg1));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_create_comment_Document(arg0, arg1, arg2) {
    let varg1 = getStringFromWasm(arg1, arg2);
    return addHeapObject(getObject(arg0).createComment(varg1));
}

export function __widl_f_create_element_Document(arg0, arg1, arg2, exnptr) {
    let varg1 = getStringFromWasm(arg1, arg2);
    try {
        return addHeapObject(getObject(arg0).createElement(varg1));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_create_range_Document(arg0, exnptr) {
    try {
        return addHeapObject(getObject(arg0).createRange());
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_create_text_node_Document(arg0, arg1, arg2) {
    let varg1 = getStringFromWasm(arg1, arg2);
    return addHeapObject(getObject(arg0).createTextNode(varg1));
}

function isLikeNone(x) {
    return x === undefined || x === null;
}

export function __widl_f_get_element_by_id_Document(arg0, arg1, arg2) {
    let varg1 = getStringFromWasm(arg1, arg2);

    const val = getObject(arg0).getElementById(varg1);
    return isLikeNone(val) ? 0 : addHeapObject(val);

}

export function __widl_f_remove_attribute_Element(arg0, arg1, arg2, exnptr) {
    let varg1 = getStringFromWasm(arg1, arg2);
    try {
        getObject(arg0).removeAttribute(varg1);
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_set_attribute_Element(arg0, arg1, arg2, arg3, arg4, exnptr) {
    let varg1 = getStringFromWasm(arg1, arg2);
    let varg3 = getStringFromWasm(arg3, arg4);
    try {
        getObject(arg0).setAttribute(varg1, varg3);
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_tag_name_Element(ret, arg0) {

    const retptr = passStringToWasm(getObject(arg0).tagName);
    const retlen = WASM_VECTOR_LEN;
    const mem = getUint32Memory();
    mem[ret / 4] = retptr;
    mem[ret / 4 + 1] = retlen;

}

export function __widl_f_inner_html_Element(ret, arg0) {

    const retptr = passStringToWasm(getObject(arg0).innerHTML);
    const retlen = WASM_VECTOR_LEN;
    const mem = getUint32Memory();
    mem[ret / 4] = retptr;
    mem[ret / 4 + 1] = retlen;

}

export function __widl_f_outer_html_Element(ret, arg0) {

    const retptr = passStringToWasm(getObject(arg0).outerHTML);
    const retlen = WASM_VECTOR_LEN;
    const mem = getUint32Memory();
    mem[ret / 4] = retptr;
    mem[ret / 4 + 1] = retlen;

}

export function __widl_f_replace_with_with_node_1_Element(arg0, arg1, exnptr) {
    try {
        getObject(arg0).replaceWith(getObject(arg1));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_instanceof_EventTarget(idx) {
    return getObject(idx) instanceof EventTarget ? 1 : 0;
}

export function __widl_f_add_event_listener_with_callback_EventTarget(arg0, arg1, arg2, arg3, exnptr) {
    let varg1 = getStringFromWasm(arg1, arg2);
    try {
        getObject(arg0).addEventListener(varg1, getObject(arg3));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_alt_HTMLImageElement(ret, arg0) {

    const retptr = passStringToWasm(getObject(arg0).alt);
    const retlen = WASM_VECTOR_LEN;
    const mem = getUint32Memory();
    mem[ret / 4] = retptr;
    mem[ret / 4 + 1] = retlen;

}

export function __widl_f_append_child_Node(arg0, arg1, exnptr) {
    try {
        return addHeapObject(getObject(arg0).appendChild(getObject(arg1)));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_contains_Node(arg0, arg1) {
    return getObject(arg0).contains(getObject(arg1));
}

export function __widl_f_is_same_node_Node(arg0, arg1) {
    return getObject(arg0).isSameNode(getObject(arg1));
}

export function __widl_f_remove_child_Node(arg0, arg1, exnptr) {
    try {
        return addHeapObject(getObject(arg0).removeChild(getObject(arg1)));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_node_type_Node(arg0) {
    return getObject(arg0).nodeType;
}

export function __widl_f_node_name_Node(ret, arg0) {

    const retptr = passStringToWasm(getObject(arg0).nodeName);
    const retlen = WASM_VECTOR_LEN;
    const mem = getUint32Memory();
    mem[ret / 4] = retptr;
    mem[ret / 4 + 1] = retlen;

}

export function __widl_f_child_nodes_Node(arg0) {
    return addHeapObject(getObject(arg0).childNodes);
}

export function __widl_f_node_value_Node(ret, arg0) {
    const val = getObject(arg0).nodeValue;
    const retptr = isLikeNone(val) ? [0, 0] : passStringToWasm(val);
    const retlen = WASM_VECTOR_LEN;
    const mem = getUint32Memory();
    mem[ret / 4] = retptr;
    mem[ret / 4 + 1] = retlen;

}

export function __widl_f_set_node_value_Node(arg0, arg1, arg2) {
    let varg1 = arg1 == 0 ? undefined : getStringFromWasm(arg1, arg2);
    getObject(arg0).nodeValue = varg1;
}

export function __widl_f_text_content_Node(ret, arg0) {
    const val = getObject(arg0).textContent;
    const retptr = isLikeNone(val) ? [0, 0] : passStringToWasm(val);
    const retlen = WASM_VECTOR_LEN;
    const mem = getUint32Memory();
    mem[ret / 4] = retptr;
    mem[ret / 4 + 1] = retlen;

}

export function __widl_f_item_NodeList(arg0, arg1) {

    const val = getObject(arg0).item(arg1);
    return isLikeNone(val) ? 0 : addHeapObject(val);

}

export function __widl_f_get_NodeList(arg0, arg1) {

    const val = getObject(arg0)[arg1];
    return isLikeNone(val) ? 0 : addHeapObject(val);

}

export function __widl_f_length_NodeList(arg0) {
    return getObject(arg0).length;
}

export function __widl_f_collapse_Range(arg0) {
    getObject(arg0).collapse();
}

export function __widl_f_set_start_Range(arg0, arg1, arg2, exnptr) {
    try {
        getObject(arg0).setStart(getObject(arg1), arg2);
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_set_start_after_Range(arg0, arg1, exnptr) {
    try {
        getObject(arg0).setStartAfter(getObject(arg1));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_set_start_before_Range(arg0, arg1, exnptr) {
    try {
        getObject(arg0).setStartBefore(getObject(arg1));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_start_container_Range(arg0, exnptr) {
    try {
        return addHeapObject(getObject(arg0).startContainer);
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_start_offset_Range(arg0, exnptr) {
    try {
        return getObject(arg0).startOffset;
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_end_container_Range(arg0, exnptr) {
    try {
        return addHeapObject(getObject(arg0).endContainer);
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_end_offset_Range(arg0, exnptr) {
    try {
        return getObject(arg0).endOffset;
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_add_range_Selection(arg0, arg1, exnptr) {
    try {
        getObject(arg0).addRange(getObject(arg1));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_get_range_at_Selection(arg0, arg1, exnptr) {
    try {
        return addHeapObject(getObject(arg0).getRangeAt(arg1));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_remove_all_ranges_Selection(arg0, exnptr) {
    try {
        getObject(arg0).removeAllRanges();
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_instanceof_Window(idx) {
    return getObject(idx) instanceof Window ? 1 : 0;
}

export function __widl_f_get_selection_Window(arg0, exnptr) {
    try {

        const val = getObject(arg0).getSelection();
        return isLikeNone(val) ? 0 : addHeapObject(val);

    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __widl_f_document_Window(arg0) {

    const val = getObject(arg0).document;
    return isLikeNone(val) ? 0 : addHeapObject(val);

}

export function __widl_f_debug_1_(arg0) {
    console.debug(getObject(arg0));
}

export function __widl_f_error_1_(arg0) {
    console.error(getObject(arg0));
}

export function __widl_f_info_1_(arg0) {
    console.info(getObject(arg0));
}

export function __widl_f_log_1_(arg0) {
    console.log(getObject(arg0));
}

export function __widl_f_warn_1_(arg0) {
    console.warn(getObject(arg0));
}

export function __wbg_newnoargs_862ffd91d0c97e88(arg0, arg1) {
    let varg0 = getStringFromWasm(arg0, arg1);
    return addHeapObject(new Function(varg0));
}

export function __wbg_call_8d65200c7bc9d2d9(arg0, arg1, exnptr) {
    try {
        return addHeapObject(getObject(arg0).call(getObject(arg1)));
    } catch (e) {
        handleError(exnptr, e);
    }
}

export function __wbindgen_object_clone_ref(idx) {
    return addHeapObject(getObject(idx));
}

function dropObject(idx) {
    if (idx < 36) return;
    heap[idx] = heap_next;
    heap_next = idx;
}

export function __wbindgen_object_drop_ref(i) { dropObject(i); }

export function __wbindgen_string_new(p, l) {
    return addHeapObject(getStringFromWasm(p, l));
}

export function __wbindgen_debug_string(i, len_ptr) {
    const toString = Object.prototype.toString;
    const debug_str = val => {
        // primitive types
        const type = typeof val;
        if (type == 'number' || type == 'boolean' || val == null) {
            return  `${val}`;
        }
        if (type == 'string') {
            return `"${val}"`;
        }
        if (type == 'symbol') {
            const description = val.description;
            if (description == null) {
                return 'Symbol';
            } else {
                return `Symbol(${description})`;
            }
        }
        if (type == 'function') {
            const name = val.name;
            if (typeof name == 'string' && name.length > 0) {
                return `Function(${name})`;
            } else {
                return 'Function';
            }
        }
        // objects
        if (Array.isArray(val)) {
            const length = val.length;
            let debug = '[';
            if (length > 0) {
                debug += debug_str(val[0]);
            }
            for(let i = 1; i < length; i++) {
                debug += ', ' + debug_str(val[i]);
            }
            debug += ']';
            return debug;
        }
        // Test for built-in
        const builtInMatches = /\[object ([^\]]+)\]/.exec(toString.call(val));
        let className;
        if (builtInMatches.length > 1) {
            className = builtInMatches[1];
        } else {
            // Failed to match the standard '[object ClassName]'
            return toString.call(val);
        }
        if (className == 'Object') {
            // we're a user defined class or Object
            // JSON.stringify avoids problems with cycles, and is generally much
            // easier than looping through ownProperties of `val`.
            try {
                return 'Object(' + JSON.stringify(val) + ')';
            } catch (_) {
                return 'Object';
            }
        }
        // errors
        if (val instanceof Error) {
        return `${val.name}: ${val.message}
        ${val.stack}`;
    }
    // TODO we could test for more things here, like `Set`s and `Map`s.
    return className;
};
const val = getObject(i);
const debug = debug_str(val);
const ptr = passStringToWasm(debug);
getUint32Memory()[len_ptr / 4] = WASM_VECTOR_LEN;
return ptr;
}

function freeCaretPosition(ptr) {

    wasm.__wbg_caretposition_free(ptr);
}
/**
* A caret position specifies the offset in the HTML source code relative to
* the start of the wrapper element.
*
* Example: If the caret is at the end of this wrapper element:
*
* ```html
* <div id=\"wrapper\">hi<br></div>
* ```
*
* ...then the offset will be 6.
*/
export class CaretPosition {

    static __wrap(ptr) {
        const obj = Object.create(CaretPosition.prototype);
        obj.ptr = ptr;

        return obj;
    }

    free() {
        const ptr = this.ptr;
        this.ptr = 0;
        freeCaretPosition(ptr);
    }

    /**
    * @returns {number}
    */
    get start() {
        return wasm.__wbg_get_caretposition_start(this.ptr);
    }
    set start(arg0) {
        return wasm.__wbg_set_caretposition_start(this.ptr, arg0);
    }
    /**
    * @returns {number}
    */
    get end() {
        return wasm.__wbg_get_caretposition_end(this.ptr);
    }
    set end(arg0) {
        return wasm.__wbg_set_caretposition_end(this.ptr, arg0);
    }
}

function freeComposeArea(ptr) {

    wasm.__wbg_composearea_free(ptr);
}
/**
* The context object containing the state.
*/
export class ComposeArea {

    static __wrap(ptr) {
        const obj = Object.create(ComposeArea.prototype);
        obj.ptr = ptr;

        return obj;
    }

    free() {
        const ptr = this.ptr;
        this.ptr = 0;
        freeComposeArea(ptr);
    }

    /**
    * Handle the specified key.
    *
    * Return whether the default keyup event handler should be prevented from running.
    * @param {string} arg0
    * @returns {boolean}
    */
    process_key(arg0) {
        const ptr0 = passStringToWasm(arg0);
        const len0 = WASM_VECTOR_LEN;
        try {
            return (wasm.composearea_process_key(this.ptr, ptr0, len0)) !== 0;

        } finally {
            wasm.__wbindgen_free(ptr0, len0 * 1);

        }

    }
    /**
    * Insert an image.
    * @param {string} arg0
    * @param {string} arg1
    * @param {string} arg2
    * @returns {void}
    */
    insert_image(arg0, arg1, arg2) {
        const ptr0 = passStringToWasm(arg0);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm(arg1);
        const len1 = WASM_VECTOR_LEN;
        const ptr2 = passStringToWasm(arg2);
        const len2 = WASM_VECTOR_LEN;
        return wasm.composearea_insert_image(this.ptr, ptr0, len0, ptr1, len1, ptr2, len2);
    }
    /**
    * Update the caret position.
    *
    * Call this after every action that might have modified the DOM.
    * @returns {void}
    */
    update_caret_position() {
        return wasm.composearea_update_caret_position(this.ptr);
    }
}

export function __wbindgen_throw(ptr, len) {
    throw new Error(getStringFromWasm(ptr, len));
}
